tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb

```diff
@@ -0,0 +1,1904 @@
+        "kl += tf.reduce_sum(qw0.kl_divergence(w0_prior))\n",
+        "kl += tf.reduce_sum(qw1.kl_divergence(w1_prior))\n",
+        "kl += tf.reduce_sum(qw2.kl_divergence(w2_prior))\n",
+        "\n",
+        "tf.summary.scalar(\"kl\", kl);"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "MuWDAFzpO4wI",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "The Log-likelihood and Kullback–Leibler divergence are then combined into the Evidence Lower BOund (ELBO) loss. We will minimize the ELBO using the Adam Algorithm ([Kingma et al., 2014](https://arxiv.org/abs/1412.6980))."
+      ]
+    },
+    {
+      "metadata": {
+        "id": "WDxJhffQq0DN",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "elbo = log_likelihood - kl\n",
+        "tf.summary.scalar(\"elbo\", elbo)\n",
+        "optimizer = tf.train.AdamOptimizer(learning_rate)\n",
+        "train_op = optimizer.minimize(-elbo)\n"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "CS9s0ntgOOPy",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "Now we can finally run our inference. \n",
+        "\n",
+        "**NOTE**: The loss function can sometimes erroneously output a negative value or `NaN`. This happens when the samples from the variational approximation are numerically zero, which causes Gamma log probs to output `inf`."
+      ]
+    },
+    {
+      "metadata": {
+        "id": "uj8JX-2NRHH2",
+        "colab_type": "code",
+        "outputId": "fb2c9cef-d311-4c0b-a145-da99cf67085f",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 9537
+        }
+      },
+      "cell_type": "code",
+      "source": [
+        "summary = tf.summary.merge_all()\n",
+        "summary_writer = tf.summary.FileWriter(model_dir, sess.graph)\n",
+        "start_time = time.time()\n",
+        "\n",
+        "evaluate(tf.global_variables_initializer())\n",
+        "for step in range(max_steps):\n",
+        "    start_time = time.time()\n",
+        "    _, elbo_value = evaluate([train_op, elbo])\n",
+        "    if step % 500 == 0:\n",
+        "        duration = time.time() - start_time\n",
+        "        print(\"Step: {:>3d} Loss: {:.3f} ({:.3f} sec)\".format(\n",
+        "              step, elbo_value, duration))\n",
+        "        summary_str = evaluate(summary)\n",
+        "        summary_writer.add_summary(summary_str, step)\n",
+        "        summary_writer.flush()\n",
+        "\n",
+        "        # Compute perplexity of the full data set. The model's negative\n",
+        "        # log-likelihood of data is upper bounded by the variational objective.\n",
+        "        negative_log_likelihood = -elbo_value\n",
```

Comment:
not much point in printing both the positive and negative likelihoods.

I think this cell needs to succeed more often than it fails to check it in. Can we lower the learning rate or adjust the prior gamma params?